opencog/rule-engine/forwardchainer/ForwardChainer.h

```diff
@@ -42,35 +49,84 @@ class ForwardChainer {
 	Handle _rbs;                // rule-based system
-
 	UREConfigReader _configReader;
-
-
     FCMemory _fcmem;            // stores history
     Logger * _log;
     int _iteration = 0;
+    source_selection_mode _ts_mode;
 
-    /**
-     * initialize config methods
-     */
     void init();
     void add_to_source_list(Handle h);
 
-    void do_pm();
-    void do_pm(const Handle& hsource, const UnorderedHandleSet& var_nodes,
-               ForwardChainerCallBack& fcb);
+    void apply_all_rules(bool search_focus_set = false);
+
+    void do_pm(const Handle& hsource, const UnorderedHandleSet& var_nodes);
+
+    UnorderedHandleSet do_step(bool search_focus_set = false);
+
+    bool is_valid_implicant(const Handle& h);
+
+    void validate(Handle hsource, HandleSeq hfocus_set);
+
+    UnorderedHandleSet get_subatoms(Rule *rule);
+
+    Handle gen_sub_varlist(const Handle& parent, const Handle& parent_varlist);
+
+    HandleSeq substitute_rule_part(
+            AtomSpace& as, Handle hrule, const std::set<Handle>& vars,
+            const std::vector<std::map<Handle, Handle>>& var_groundings);
+
+    bool unify(Handle source, Handle target, Rule* rule);
+    bool subatom_unify(Handle source, Rule* rule);
+    HandleSeq derive_rules(Handle source, Handle target, Rule* rule);
+
 protected:
-    enum source_selection_mode {
-        TV_FITNESS_BASED, STI_BASED
-    };
+    /**
+     * Choose an applicable rules from the rule base by selecting
+     * rules whose premise structurally matches with the source.
+     *
+     * @return  A rule that in which @param source could ground.
+     */
+    virtual Rule* choose_rule(Handle hsource, bool subatom_match );
+
+    /**
+     * Choose additional premises for the rule.
+     *
+     * @fcmem  An object holding the current source/target and other inform
+     *         ation of the forward chaining instance.
+     *
+     * @return  A set of Handles chosen as a result of applying fitness
+     *          criteria with respect to the current source.
+     */
+
+    virtual HandleSeq choose_premises(FCMemory& fcmem);
+    /**
+     * choose next source from the source list
+     *
+     * @return  A handle to the chosen source from source list
+     */
+    virtual Handle choose_next_source(FCMemory& fcmem);
+
+    /**
+     * Apply chosen rule. the default will wrap a custom PM callback class.
+     * i.e invokes _pattern_matcher.
+     *
+     * @return  A set of handles created as a result of applying current
+     *          choosen rule.
+     */
+    virtual HandleSeq apply_rule(Handle rhandle, bool search_focus_set_only =
+            false);
+
+    HandleSeq derive_rules(Handle source, Rule* rule, bool subatomic = false);
+
 public:
 	/**
 	 * Ctor. rbs is a Handle pointing to rule-based system.
 	 */
     ForwardChainer(AtomSpace& as, Handle rbs);
-    void do_chain(ForwardChainerCallBack& fcb, Handle hsource =
-            Handle::UNDEFINED);
-    void do_step(ForwardChainerCallBack& fcb);
```

Comment:
I wonder about the removal of do_step...  it means it is impossible to implement a combined FC/BC (I envision do_step will be used to control one step of FC and one step of BC at a time).